ceph/rados/mgr_workflows.py

```diff
@@ -0,0 +1,274 @@
+from ceph.rados.monitor_workflows import MonitorWorkflows
+from utility.log import Log
+
+log = Log(__name__)
+
+
+class MgrWorkflows:
+    """
+    Module to perform Serviceability scenarios on mgr daemons
+    """
+
+    def __init__(self, node: CephAdmin):
+        """
+        Module to perform Serviceability scenarios on mgr daemons
+
+        Args:
+            node: CephAdmin object
+        """
+        self.rados_obj = RadosOrchestrator(node=node)
+        self.mon_obj = MonitorWorkflows(node=node)
+        self.cluster = node.cluster
+        self.config = node.config
+        self.client = node.cluster.get_nodes(role="client")[0]
+        self.cluster_nodes = node.cluster.get_nodes()
+
+    def add_mgr_service(self, **kwargs) -> bool:
+        """
+        Method to add the mgr service from the given host
+        Args:
+            kwargs: the accepted KW args for the method are:
+                host: Cluster object of the host from where mgr needs to be Added
+        returns:
+            Pass -> True, Fail -> false
+        """
+        # Checking if mgr service exists before removing
+        host = kwargs.get("host")
+        hostname = host.hostname
+
+        if self.check_mgr_exists_on_host(host=hostname):
+            log.info(
+                f"MGR daemon already present on the host {hostname}. Returning Pass"
+            )
+            return True
+
+        log.debug(f"Adding mgr daemon on the host : {hostname}")
+
+        # Adding lable to host
+        labels = self.mon_obj.get_host_labels(host=hostname)
+        if "mgr" not in labels:
+            cmd = f"ceph orch host label add {hostname} mgr"
+            self.client.exec_command(sudo=True, cmd=cmd)
+            time.sleep(10)
+            log.debug(f"Added mgr label for the host : {hostname}")
+
+        # Adding host as mgr
+        cmd = "ceph orch apply mgr --placement='label:mgr'"
+        self.client.exec_command(sudo=True, cmd=cmd)
+        time.sleep(10)
+
+        endtime = datetime.datetime.now() + datetime.timedelta(seconds=180)
+        while datetime.datetime.now() < endtime:
+            if self.check_mgr_exists_on_host(host=hostname):
+                log.info(f"Mgr {hostname} is add to the cluster.")
+                return True
+            log.info(
+                "Mgr daemon still not added.sleeping for 5 seconds and checking again"
+            )
+            time.sleep(5)
+        return False
+
+    def remove_mgr_service(self, host) -> bool:
+        """
+        Method to remove the mgr service from the given host
+
+        Args:
+            host: name of the host from where mgr needs to be removed
+
+        returns:
+            Pass -> True, Fail -> false
+        """
+        mgr_daemon_list = self.get_mgr_daemon_list()
+        new_mgr_daemon = ""
+        for node in mgr_daemon_list:
+            if host.hostname in node:
+                new_mgr_daemon = node.strip()
+                break
+
```

Comment:
What is the purpose of this code block?

Reply:
This block is to get the exact mgr name. To remove mgr from the node it needs to provide the mgr node name.
Example: From the below the mgr node names are - ceph-rados-auto-65guw8-node2.zgoyni, ceph-rados-auto-65guw8-node1-installer.welqwn, ceph-rados-auto-65guw8-node3.xgkqzi and ceph-rados-auto-65guw8-node6.sjgsmd


[root@ceph-rados-auto-65guw8-node7 cephuser]# ceph -s
  cluster:
    id:     5400db6e-224e-11ef-8b24-fa163e595ce1
    health: HEALTH_OK
 
  services:
    mon: 5 daemons, quorum ceph-rados-auto-65guw8-node1-installer,ceph-rados-auto-65guw8-node3,ceph-rados-auto-65guw8-node4,ceph-rados-auto-65guw8-node2,ceph-rados-auto-65guw8-node6 (age 2h)
    mgr: ceph-rados-auto-65guw8-node2.zgoyni(active, since 12m), standbys: ceph-rados-auto-65guw8-node1-installer.welqwn, ceph-rados-auto-65guw8-node3.xgkqzi, ceph-rados-auto-65guw8-node6.sjgsmd
    osd: 15 osds: 15 up (since 2h), 15 in (since 2h)
 
  data:
    pools:   1 pools, 1 pgs
    objects: 2 objects, 449 KiB
    usage:   405 MiB used, 375 GiB / 375 GiB avail
    pgs:     1 active+clean
 
[root@ceph-rados-auto-65guw8-node7 cephuser]#